### PR TITLE
usxxx internal user sign in with uaa

### PR DESCRIPTION
**Pull request by JamesGardiner**
This PR makes the use of UAA configurable via the app's environment. If USE_UAA is set in the app's environment then the app will attempt to use UAA to authenticate. This option is only in place to allow local dev to continue whilst a local UAA mock/docker solution is in place. In future iterations UAA will be the only authentication mechanism.

### CODE REVIEW

File: ras_backstage/controllers/uaa_controller.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+from ras_backstage.common.requests_handler import request_handler
+from ras_backstage.exception.exceptions import ApiError
+
+
+logger = wrap_logger(logging.getLogger(__name__))
+
+
+def sign_in(username, password):
+    logger.debug('Retrieving OAuth2 token for sign-in')
+    url = '{}{}'.format(app.config['UAA_SERVICE_URL'], '/oauth/token')
```
Comment (dcdarrell9) [inline ras_backstage/controllers/uaa_controller.py]: Shall we use an f'string here for consistency?

File: config.py [hunk 2]
```diff
@@ -76,6 +76,11 @@ class Config(object):
                                              RM_SAMPLE_SERVICE_HOST,
                                              RM_SAMPLE_SERVICE_PORT)
 
+    UAA_SERVICE_URL = os.getenv('UAA_SERVICE_URL', 'localhost')
+    UAA_CLIENT_ID = os.getenv('UAA_CLIENT_ID', 'ras_backstage_client_id')
+    UAA_CLIENT_SECRET = os.getenv('UAA_CLIENT_SECRET', 'password')
+    USE_UAA = os.getenv('USE_UAA', True)
```
Comment (jcox-dev) [inline config.py]: Will this ever be False?

File: ras_backstage/resources/sign_in/sign_in.py [hunk 3]
```diff
@@ -53,12 +53,28 @@ def post():
...
 
-        # Obviously horrible, stopgap until uaa is implemented
-        if username == current_app.config['USERNAME'] and password == current_app.config['PASSWORD']:
-            # We're assuming that uaa will return an Oauth2 token though it's almost certain that
-            # this will change once we know exactly what is being returned.
-            logger.info("Authentication successful", user=username)
-            return make_response(jsonify({"token": "1234abc"}), 201)
+        if not current_app.config.get('USE_UAA'):
+            logger.info('Retrieving sign-in details')
+            if username == current_app.config['USERNAME'] and password == current_app.config['PASSWORD']:
```
Comment (jcox-dev) [inline ras_backstage/resources/sign_in/sign_in.py]: If no username and password is provided and we're not using the dev config would this evaluate to True (`None == None and None == None`) if the config isn't set correctly?

Comment (dcdarrell9) [reply ras_backstage/resources/sign_in/sign_in.py]: There's validation to make sure a user cannot send no username or password, but might be worth having a test for something like this 👍

Comment (insacuri) [reply ras_backstage/resources/sign_in/sign_in.py]: Form validation kicks in before you get to this point which stops you from submitting nothing

File: config.py (hunk 2 above)
Comment (JamesGardiner) [reply config.py]: Ah good catch. Happy to wrap it in a `bool()`?

File: ras_backstage/resources/sign_in/sign_in.py (hunk 3 above)
Comment (JamesGardiner) [reply ras_backstage/resources/sign_in/sign_in.py]: @jcox-dev yes

File: ras_backstage/resources/sign_in/sign_in.py [hunk 4]
```diff
@@ -53,12 +53,28 @@ def post():
...
-            logger.info("Authentication failed", user=username)
-            return make_response(jsonify({"error": "Username and/or password incorrect"}), 401)
+            logger.info('Retrieving sign-in details using UAA')
+            message_json = request.get_json()
+            username = message_json.get('username')
+            password = message_json.get('password')
+
+            oauth2_token = uaa_controller.sign_in(username, password)
+            token_expiry = datetime.now() + timedelta(seconds=int(oauth2_token['expires_in']))
+            oauth2_token['expires_at'] = token_expiry.timestamp()
```
Comment (andrewmil) [inline ras_backstage/resources/sign_in/sign_in.py]: Can we set an expiry time in the uaa service or do we have to set it here?

File: ras_backstage/resources/sign_in/sign_in.py [hunk 5]
```diff
@@ -53,12 +53,28 @@ def post():
...
+            if username == current_app.config['USERNAME'] and password == current_app.config['PASSWORD']:
+                logger.info("Authentication successful", user=username)
+                return make_response(jsonify({"token": "1234abc"}), 201)
+            else:
+                logger.info("Authentication failed", user=username)
+                return make_response(jsonify({"error": "Username and/or password incorrect"}), 401)
         else:
-            logger.info("Authentication failed", user=username)
-            return make_response(jsonify({"error": "Username and/or password incorrect"}), 401)
+            logger.info('Retrieving sign-in details using UAA')
```
Comment (andrewmil) [inline ras_backstage/resources/sign_in/sign_in.py]: We should probably add tests for signing in with UAA

File: test/resources/test_collection_instrument.py [hunk 6]
```diff
@@ -39,14 +37,9 @@ def setUp(self):
...
     def test_upload_collection_instrument(self, mock_request):
-        classifiers = {
-            "SURVEY_ID": "cb0711c3-0ac8-41d3-ae0e-567e5ea1ef87",
-            "COLLECTION_EXERCISE": "e33daf0e-6a27-40cd-98dc-c6231f50e84a"
-        }
-        params = urlencode({'classifiers': json.dumps(classifiers)})
         mock_request.get(url_get_survey_by_short_name, json=self.survey)
         mock_request.get(url_ces, json=self.collection_exercises)
-        mock_request.post(f'{url_upload_collection_instrument}?{params}', complete_qs=True,
+        mock_request.post(f'{url_upload_collection_instrument}', complete_qs=True,
```
Comment (andrewmil) [inline test/resources/test_collection_instrument.py]: If we're removing `?{params}` here then we no longer need `complete_qs=True`

File: ras_backstage/resources/sign_in/sign_in.py (hunk 4 above)
Comment (warrenbailey) [reply ras_backstage/resources/sign_in/sign_in.py]: None of this is actually needed, we're going to use the token directly from the UAA

**Comment by JamesGardiner** [comment]
Closing pull request for now - UAA changes will mean significant changes to the code in this branch. Will reopen once uaa implemented and we have a stable interface.

**Pull request closed without merging**